iquip/apps/builder.py

```diff
@@ -309,11 +312,83 @@ def __init__(
         """
         super().__init__(parent=parent)
         self.name = name
-        self.argInfo = argInfo
+        self.state = self.get_state(argInfo)
+        self.stack = QStackedWidget(self)
+        self.layout = QGridLayout(self)
+        self.layout.addWidget(QLabel(name, self), 0, 0)
+        self.layout.addWidget(self.stack, 1, 1)
+
+    def get_state(self, argInfo):
+        """Gets a state dictionary that describes key parameters of a scannable object.
+
+        Creates a default state dictionary and updates it using the argInfo.
+
+        Args:
+            argInfo: A dictionary that contains argument information of each scannable type.
+        """
+        scale = argInfo["scale"]
+        state = {
+            "selected": "NoScan",
+            "NoScan": {"value": 0.0, "repetitions": 1},
+            "RangeScan": {"start": 0.0, "stop": 100.0*scale, "npoints": 10,
+                          "randomize": False, "seed": None},
+            "CenterScan": {"center": 0.*scale, "span": 100.*scale,
+                           "step": 10.*scale, "randomize": False,
+                           "seed": None},
+            "ExplicitScan": {"sequence": []}
+        }
+        if "default" in argInfo:
+            defaults = argInfo["default"]
+            if not isinstance(defaults, list):
+                defaults = [defaults]
+            state["selected"] = defaults[0]["ty"]
+            for default in reversed(defaults):
+                ty = default["ty"]
+                if ty == "NoScan":
+                    state[ty]["value"] = default["value"]
+                    state[ty]["repetitions"] = default["repetitions"]
+                elif ty == "RangeScan":
+                    state[ty]["start"] = default["start"]
+                    state[ty]["stop"] = default["stop"]
+                    state[ty]["npoints"] = default["npoints"]
+                    state[ty]["randomize"] = default["randomize"]
+                    state[ty]["seed"] = default["seed"]
+                elif ty == "CenterScan":
+                    for key in "center span step randomize seed".split():
+                        state[ty][key] = default[key]
+                elif ty == "ExplicitScan":
+                    state[ty]["sequence"] = default["sequence"]
+                else:
+                    logger.warning("unknown default type: %s", ty)
+        return state
+
+    def get_procdesc(self, argInfo):
+        """Gets a prodesc dictionary that describes common parameters of the scannable object.
+
+        Args:
+            argInfo: A dictionary that contains argument information of the scannable object.
+        """
+        procdesc = {
+            "unit": argInfo["unit"],
+            "scale": argInfo["scale"],
+            "global_step": argInfo["global_step"],
+            "ndecimals": argInfo["ndecimals"],
+            "global_min": argInfo["global_min"],
+            "global_max":  argInfo["global_max"]
+        }
+        return procdesc
```

Comment:
I think you can return this dict directly, not defining a variable.
However it is just a coding style, do as you please.

Reply:
I will reconsider it at the next PR.